src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs

```diff
@@ -543,145 +545,131 @@ private async ValueTask FinishBatchOfDocumentsAsync()
 
         private class DatabaseCompareExchangeActions : ICompareExchangeActions
         {
+            const int BatchSize = 1024;
+
             private readonly DocumentDatabase _database;
             private readonly JsonOperationContext _context;
             private readonly List<RemoveCompareExchangeCommand> _compareExchangeRemoveCommands = new List<RemoveCompareExchangeCommand>();
             private readonly List<AddOrUpdateCompareExchangeCommand> _compareExchangeAddOrUpdateCommands = new List<AddOrUpdateCompareExchangeCommand>();
+            private DisposableReturnedArray<BatchRequestParser.CommandData> _commandParsedCommands = new DisposableReturnedArray<BatchRequestParser.CommandData>(1024);
+            private (IDisposable disposable, DocumentsOperationContext ctx, DocumentsTransaction readTrx) _operationContext;
 
             public DatabaseCompareExchangeActions(DocumentDatabase database, JsonOperationContext context)
             {
                 _database = database;
                 _context = context;
+                _operationContext.disposable = _database.DocumentsStorage.ContextPool.AllocateOperationContext(out _operationContext.ctx);
+            }
+
+            struct DisposableReturnedArray<T> : IDisposable
+            {
+                private readonly T[] _array;
+                
+                public int Length;
+
+                public DisposableReturnedArray(int length)
+                {
+                    _array = ArrayPool<T>.Shared.Rent(length);
+                    Length = 0;
+                }
+
+                public void Push(T item) => _array[Length++] = item;
+                public T this[int index] => _array[index];
+
+                public ArraySegment<T> GetArraySegment() => new ArraySegment<T>(_array, 0, Length);
+
+                public void Clear() => Length = 0; 
+
+                public void Dispose() => ArrayPool<T>.Shared.Return(_array);
             }
 
             public async ValueTask WriteKeyValueAsync(string key, BlittableJsonReaderObject value)
             {
-                const int batchSize = 1024;
+                if(_operationContext.readTrx.Disposed)
+                    _operationContext.readTrx = _operationContext.ctx.OpenReadTransaction();
+
                 _compareExchangeAddOrUpdateCommands.Add(new AddOrUpdateCompareExchangeCommand(_database.Name, key, value, 0, _context, RaftIdGenerator.DontCareId, fromBackup: true));
+                if (ClusterTransactionCommand.IsAtomicGuardKey(key, out var docId))
```

Comment:
Note that we don't want to add the compare exchange here if it is an atomic guard.